ipld/get_namespaced_shares.go

```diff
@@ -0,0 +1,226 @@
+package ipld
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/celestiaorg/nmt"
+
+	"go.opentelemetry.io/otel/trace"
+
+	"go.opentelemetry.io/otel/attribute"
+
+	"github.com/gammazero/workerpool"
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+
+	"github.com/celestiaorg/celestia-node/ipld/plugin"
+
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+// TODO(@distractedm1nd) Find a better figure than NumWorkersLimit for this pool.
+// Worker pool responsible for the goroutines spawned by getLeavesByNamespace
+var namespacePool = workerpool.New(NumWorkersLimit)
+
+// GetSharesByNamespace walks the tree of a given root and returns its shares within the given namespace.ID.
+// If a share could not be retrieved, err is not nil, and the returned array contains nil shares.
+func GetSharesByNamespace(
+	ctx context.Context,
+	bGetter blockservice.BlockGetter,
+	root cid.Cid,
+	nID namespace.ID,
+	maxShares int,
+) ([]Share, error) {
+	ctx, span := tracer.Start(ctx, "get-shares-by-namespace")
+	defer span.End()
+
+	leaves, err := getLeavesByNamespace(ctx, bGetter, root, nID, maxShares)
+	if err != nil && leaves == nil {
+		return nil, err
+	}
+
+	shares := make([]Share, len(leaves))
+	for i, leaf := range leaves {
+		if leaf != nil {
+			shares[i] = leafToShare(leaf)
+		}
+	}
+
+	return shares, err
+}
+
+// wrappedWaitGroup is needed because waitgroups do not expose their internal counter,
+// and we don't know in advance how many jobs we will have to wait for.
+type wrappedWaitGroup struct {
+	wg      sync.WaitGroup
+	jobs    chan *job
+	counter int64
+}
+
+func (w *wrappedWaitGroup) Add(count int64) {
+	w.wg.Add(int(count))
+	atomic.AddInt64(&w.counter, count)
+}
+
+func (w *wrappedWaitGroup) Done() {
+	w.wg.Done()
+	numRemaining := atomic.AddInt64(&w.counter, -1)
+
+	// Close channel if this job was the last one
+	if numRemaining == 0 {
+		close(w.jobs)
+	}
+}
+
+func (w *wrappedWaitGroup) Wait() {
+	w.wg.Wait()
+}
+
+type fetchedBounds struct {
```

Comment:
Optional refactoring: Extending this struct with the actual nodes/leaves seems reasonable. E.g. when you do `Update`, you can also pass the node as well. And then, It will also have a method `Result`(or else) that encapsulates slicing.

Reply:
I feel like keeping the "separation of concerns" principle is better here, so I'll keep it as it is, but I'm on the fence about this suggestion - both solutions are nice in their own ways